src/printer/model_library_format_printer.cc

```diff
@@ -0,0 +1,75 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+#include <tvm/runtime/module.h>
+#include <tvm/runtime/registry.h>
+#include <tvm/tir/var.h>
+
+#include "text_printer.h"
+
+namespace tvm {
+namespace printer {
+
+class ModelLibraryFormatPrinter : public ::tvm::runtime::ModuleNode {
```

Comment:
Why are we extending runtime.Module ? 
Any reason why we cant use Objects and Nodes to expose this to python?

Reply:
we need `GetFunction` to expose an object to Python with member functions, and i think the member function style provides a bit more structure to the API than just e.g. placing it all in a tuple. I don't think we can use Node since it doesn't provide the `sptr_to_self`

Reply:
I was actually referring to other objects with member functions.
E.g. : 
https://github.com/manupa-arm/incubator-tvm/blob/master/src/relay/analysis/call_graph.cc 
https://github.com/manupa-arm/incubator-tvm/blob/master/python/tvm/relay/analysis/call_graph.py

There are similiar structure in AutoScheduler as well. I always this was better than extending runtime.Modules and using packed functions. What do you think?

Reply:
hmm. i think that this way is a more codified dispatch mechanism (e.g. string table lookup) than that used by Module (`GetFunction`  typically implemented with a series of if statements and closures), but it requires additional duplication of each member function in e.g. Python. I think a proper interface-oriented FFI would just wrap all of this stuff automatically.

my preference with interface FFI functions in TVM is to use `Module` in general, but add some automation to the build to auto-generate `GetFunction` and avoid closures in that function. It is a bit weird, I'll admit, since it re-uses a mechanism meant for the runtime. But, I think it's the only generic member-function lookup we have right now, and there are some non-runtime use cases.

Reply:
Yes, I think there is precedence for both approaches. Lets not block this based on this :). 
Maybe its better to converge to a single policy when exposing member functions across the FFI. 
cc : @tqchen @jroesch .